drivers: introduce memory management driver

dcpleung opened the pull request:
This introduces new APIs for memory management hardware which are not traditionally associated with an architecture (e.g. MMU on ARM and x86) and thus not implemented in architecture code. A sample driver is also added for the Intel Audio DSP to demonstrate how to use the API.

Commit:
test: boards/intel_adsp: Add test of adsp_mem API

Add a simple board-specific test of the memory mapping driver.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: arch: no need for #ifdef MMU in header

There is no need to use conditional compilation for the function
prototypes in the kernel architecture header file. So remove it.
Added bouns is that these functions can appear in documentation
without explicitly enabled in pre-defines during doc build.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: introduce memory management driver APIs

This introduces new APIs for memory management hardware which
are not traditionally associated with an architecture (e.g.
MMU on ARM and x86) and thus not implemented in architecture
code.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: move CONFIG_MMU into kernel Kconfig

This moves CONFIG_MMU and its children from arch/Kconfig into
kernel/Kconfig. These are used to enable kernel support of MMU
so they should be under kernel/.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: introduce hidden CONFIG_KERNEL_VM_SUPPORT

Introduce a hidden kconfig CONFIG_KERNEL_VM_SUPPORT which
enables some kconfigs that are required for virtual memory
support. CONFIG_KERNEL_VM_BASE, CONFIG_KERNEL_VM_OFFSET,
and CONFIG_KERNEL_VM_SIZE are moved under this new kconfig
so these can be enabled independent of CONFIG_MMU.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: mm: add skeleton build files and common funcs

This adds skeleton Kconfig/CMakeLists.txt and common implementation
of some sys_mm_drv_*() functions.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: mm: add driver for Intel Audio DSP

This adds the driver for the memory mapping hardware
on Intel Audio DSP.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

MaureenHelm commented on drivers/mem_manage/mm_drv_common.c:
```diff
@@ -290,30 +290,26 @@ int sys_mm_drv_move_region(void *virt_old, size_t size, void *virt_new,
-			 * with unknown random flags.
+			 * Only map the new page when we can retrieve
+			 * flags of the old mapped page as We don't
+			 * want to map with unknown random flags.
```
Are the changes to `drivers/mem_manage/mm_drv_common.c` in commit `drivers: mem_manage: add driver for Intel Audio DSP` a rebasing error?

dcpleung replied:
I was at the wrong commit when refreshing using stgit. Fixed.

MaureenHelm commented on drivers/mem_manage/mm_drv_intel_adsp_tlb.c:
```diff
@@ -0,0 +1,194 @@
+
+#include "mm_drv_common.h"
+
+#define TLB_BASE 0x3000
```
This should come from devicetree

dcpleung replied:
Updated to use devicetree.

However, this doesn't expose APIs as in, for example, GPIO drivers, where there are driver pointers. This is modeled similar to timer driver where there can only be one instance. Let me know where I can put it, if not in `drivers/`.

MaureenHelm reviewed (commented):
This shouldn't live in `drivers/` if it doesn't follow the zephyr device driver model.

dcpleung committed:
dts: add base binding for memory management drivers

This adds a simple base binding for memory management drivers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
dts: intel_cavs25: add an entry for the TLB driver

This adds an entry for the TLB memory management driver
on intel_cavs25.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_adsp/cavs_v25: set CONFIG_KERNEL_VM_SIZE

This sets CONFIG_KERNEL_VM_SIZE to define the size of virtual
memory region which can be used by the TLB driver. The size is
derived from the actual SOF code where they say the TLB register
space is of size 0x1000. There are 2048 TLB entries (for 2 bytes
per entry), which translates to 8MB of virtual memory space with
4KB pages.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

carlescufi commented:
@carlocaione could you review this one please?

carlocaione commented on drivers/mem_manage/mm_drv_common.c:
```diff
@@ -0,0 +1,409 @@
+
+	for (idx = 0; idx < cnt; idx++) {
+		if (!sys_mm_drv_is_addr_aligned(addr[idx])) {
+			ret = false;
+			break;
```
just `return false` and you can remove the `ret` variable.

dcpleung replied:
MISRA wants only one return statement per function... (errr...)

carlocaione commented on drivers/mem_manage/mm_drv_common.c:
```diff
@@ -0,0 +1,409 @@
+		uint8_t *va = (uint8_t *)virt + offset;
+		uintptr_t pa = phys + offset;
+
+		int ret2 = sys_mm_drv_map_page(va, pa, flags);
+
+		if (ret2 != 0) {
+			ret = ret2;
```
in case we fail to map one page we keep going and we map all the remaining and return the error. But now I guess the user cannot use that page at that `va` and he doesn't even know which page failed to be mapped. How is this supposed to be used?

dcpleung replied:
I thought of using `CHECKIF()` but decided against it as you can disable the check with kconfig, but I forgot to add back the assertion.

As for what you have stated, I don't have a good solution as we cannot simply revert changes made before the failure because there is no history saved. Returning a failed list would probably require a few malloc() which I would like to avoid. I could force a kernel panic here if you want. Note that this provides a very primitive implementation where an app specific solution can override this.

teburd replied:
You could keep history on the stack with a recursive call here, but that's probably evil as you don't know the stack size of the calling thread and are likely to run it over very quickly.

carlocaione commented on drivers/mem_manage/mm_drv_common.h:
```diff
@@ -0,0 +1,277 @@
+ */
+static inline bool sys_mm_drv_is_addr_aligned(uintptr_t addr)
+{
+	if ((addr & (CONFIG_MM_DRV_PAGE_SIZE - 1)) == 0U) {
+		return true;
+	} else {
+		return false;
+	}
```
`return ((addr & ...` ?

nashif commented on include/drivers/mem_manage/system_mem_manage.h:
```diff
@@ -0,0 +1,306 @@
+
+/**
+ * @file
+ * @brief Memory Management Driver APIs
```
How about putting this in `include/drivers/mm/` and calling it `system_mm.h` instead?

nashif commented on include/drivers/mem_manage/system_mem_manage.h:
```diff
@@ -0,0 +1,306 @@
+ * driver. Only one instance is permitted on the system.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_SYSTEM_MM_H_
```
.... which will correspond to this...

nashif commented on dts/bindings/mem_manage/mm_drv.yaml:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2021 Intel Corporation
```
dts/bindings/mm/mm_drv.yaml

nashif commented on drivers/mem_manage/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
```
drivers/mm

nashif commented on tests/boards/intel_adsp/mem/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
put it in  `tests/boards/intel_adsp/mm`

nashif reviewed (changes requested):
trivial naming changes...

dcpleung commented:
Renamed.

teburd reviewed (approved):
I looked over the API, docs and tests, they all seem good to me.
